### ISSUE TITLE
Round-robin mailserver retries

### ISSUE THREAD

**Issue by spud** [schweikert/postgrey#49]
I've been watching my mail server logs (endless fun!), and notice that with many large-scale mail hosts, the "retry" attempts come from a different hostname than the original attempt, so they ended up getting re-greylisted instead of passing the mail through.

For example, an initial connection attempt may come from

o1.sg.crunchbase.com (167.89.93.77)

but retries come from 

o2.sg.crunchbase.com (167.89.104.98)

So it appears that BOTH o1.sg.crunchbase.com and o2.sg.crunchbase.com get greylisted, and it won't be until one of them connects again after 5 minutes that the mail will pass through.

This means that mail can be delayed by (5min * [number of round robin servers in use]).

Am I interpreting this correctly, or is postgrey doing something to accommodate this situation?

**Comment by nicorac** [comment]
I'm experiencing the same issue wito outlook.com.
Some messages were delivered with up to 24 hours delay from a large mail host like outlook.com.

This is only a sample of the client hosts I've seen for a single message (that was delayed fo more than 24 hours):
```
mail-db5eur01on0044.outbound.protection.outlook.com
mail-eopbgr10051.outbound.protection.outlook.com
mail-eopbgr10087.outbound.protection.outlook.com
mail-eopbgr20064.outbound.protection.outlook.com
mail-eopbgr30079.outbound.protection.outlook.com
mail-he1eur01on0047.outbound.protection.outlook.com
mail-ve1eur01on0079.outbound.protection.outlook.com
mail-ve1eur01on0081.outbound.protection.outlook.com
```
There should be some kind of **domain depth limit** so all of them will become **outlook.com** or **protection.outlook.com**.

Meanwhile I've set a workaround in Postfix by prepending a check_sender_access option to postgrey in recipient restrictions:
```
smtpd_recipient_restrictions =
  ...
  check_sender_access hash:/etc/postfix/client-whitelist.list,  # <-- CUSTOM POSTGREY WHITELIST
  check_policy_service inet:127.0.0.1:10023,
  ...
```
but this is cumbersome and error prone.

**Comment by schweikert** [comment]
You probably should whitelist crunchbase.com in postgrey. Also, whitelisting in Postfix is also a good solution, especially if you encounter trouble with Postrey < 1.37

**Comment by spud** [comment]
Disappointing resolution. The problem isn't specific to crunchbase.com. That was merely an example. As noted by nicorac above, the problem easily occurs with outlook.com, and in my experience google.com and other mass-volume email hosts. Whitelisting each one (as they are discovered) just means that I have to deal with an irate client each time it occurs, until I've caught them all.

Is there no way for the tuple to use a pattern like

*.outbound.protection.outlook.com

instead of each individual host name?

**Comment by nicorac** [comment]
I think that we should only consider the latest 3 parts of the domain name (or better consider the last 4 to accomplish with `*.co.uk` and similar, there are public lists available for that).

There's no advantage in considering differences in levels from fourth up (see my post below).
To me `mail1.domain.com` and `mail2.domain.com` should be considered as "a message sent from `domain.com` domain.

I'm actually thinking about disabling greylisting at all, till this gets sorted out; it's giving more headaches than advantages.

**Comment by schweikert** [comment]
outlook.com is already whitelisted. It probably doesn't work because of #32, which was only solved in postgrey 1.37.
